### PR TITLE
Upgrade wallet-factory and zoe in handler

### LINKED ISSUE #8826: can't build bundles compatible with mainnet due to endo patterns breaking change
## Describe the bug

I ran into `SwingSet: xsnap: v11: RangeError#1: Expected undefined is same as Interface`, asked for help, and learned in discussion with @Chris-Hibbert, @kriskowal , @mhofman   that it's due to a breaking change in @endo/patterns, and that as a result, I wouldn't be able to test the upgrade fix for #8311 on master; I'd have to do development in the release branch, which (presumably) doesn't include the breaking change to patterns.

> 1.0.0 (2023-12-12)
> ⚠ BREAKING CHANGES
> patterns: tag and retype guards
-- [patterns/CHANGELOG](https://github.com/endojs/endo/blob/master/packages/patterns/CHANGELOG.md)

## To Reproduce

1. check out [8826-wip-repro](https://github.com/Agoric/agoric-sdk/tree/8826-wip-repro) (`683733d` as of this writing)
2. `cd agoric-sdk; yarn`
3. `cd a3p-integration; corepack enable; yarn`
4. `yarn test`
   - note: `yarn test` needs **docker** as usual for working with [agoric-3-proposals](https://github.com/Agoric/agoric-3-proposals))
6. lots of stuff happens that's cached, up thru `=> CACHED [test-names-by-addr test-names-by-addr 4/4]`
7. `Running test image for proposal upgrade-14` runs ok
8. `Running test image for proposal names-by-addr` starts; goes ok thru `v1: upgradeProvisioningVat...`, but then
9. fail with `v1: Error: vat-upgrade failure`

<details><summary>stack for v1: Error: vat-upgrade failure</summary>

```
2024-01-26T22:32:05.111Z SwingSet: vat: v1: evaluateBundleCap { manifestBundleRef: { bundleID: 'b1-ae1949442574923fc03f229d529644ded41a2dbdc6336f6bf1496fdd7f557b1fcad2d1bfa59a2f0a2241f66a8f864dcbbeddd3abf0afcf0ac7879d26d0821994' }, manifestGetterName: 'getManifestForProvisioning', vatAdminSvc: Promise [Promise] {} }
2024-01-26T22:32:05.171Z SwingSet: vat: v1: execute { manifestGetterName: 'getManifestForProvisioning', bundleExports: [ 'getManifestForProvisioning', 'upgradeProvisioningVat' ] }
2024-01-26T22:32:05.171Z SwingSet: vat: v1: coreProposal: upgradeProvisioningVat
2024-01-26T22:32:05.171Z SwingSet: vat: v1: upgradeProvisioningVat...
2024-01-26T22:32:05.172Z SwingSet: vat: v1: provisioning { adminNode: Object [Alleged: adminNode] {}, root: Object [Alleged: undefined] {} } b1-10d640ccc3f09d37719d87bf33b4cf2cf950da2a5daf7b2c927e5b4569e65e03950fa36be3591e33250906a98e6c39f79fd4cfb5fe1b38f5aff8a115ccfe3846
2024-01-26T22:32:05.174Z SwingSet: vat: v1: CORE_EVAL failed: (RemoteError(error:liveSlots:v2#70001)#3)
2024-01-26T22:32:05.174Z SwingSet: vat: v1: RemoteError(error:liveSlots:v2#70001)#3: vat-upgrade failure
2024-01-26T22:32:05.174Z SwingSet: vat: v1: Error: vat-upgrade failure
 at construct ()
 at Error (/bundled-source/.../node_modules/ses/src/error/tame-error-constructor.js:56)
 at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:243)
 at decodeErrorCommon (/bundled-source/.../node_modules/@endo/marshal/src/marshal.js:281)
 at decodeFromSmallcaps (/bundled-source/.../node_modules/@endo/marshal/src/encodeToSmallcaps.js:434)
 at fromCapData (/bundled-source/.../node_modules/@endo/marshal/src/marshal.js:356)
 at notifyOnePromise (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1225)
 at notify (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1249)
 at dispatchToUserspace (/bundled-source/.../packages/swingset-liveslots/src/liveslots.js:1514)
 at runWithoutMetering (/bundled-source/.../packages/swingset-xsnap-supervisor/lib/supervisor-subprocess-xsnap.js:60)
 at ()
```
</details> 

## Expected behavior

The upgrade should work just as well as on the release branch (#8786).

## Platform Environment

_IOU_

 - what OS are you using? what version of Node.js?
 - is there anything special/unusual about your platform?
 - what version of the Agoric-SDK are you using? (run `git describe --tags --always`)

## Additional context

was working on #8311

## Screenshots / Stack trace

<details><summary>earlier stack from 2024-01-22</summary>

```
ag-power-tools-agd-1  | 2024-01-22T17:19:40.483Z SwingSet: vat: v1: provisioning { adminNode: Object [Alleged: adminNode] {}, root: Object [Alleged: undefined] {} } b1-10d640ccc3f09d37719d87bf33b4cf2cf950da2a5daf7b2c927e5b4569e65e03950fa36be3591e33250906a98e6c39f79fd4cfb5fe1b38f5aff8a115ccfe3846
ag-power-tools-agd-1  | 2024-01-22T17:19:41.117Z SwingSet: xsnap: v11: error during vat dispatch() of startVat,[object Object] (RangeError#1)
ag-power-tools-agd-1  | 2024-01-22T17:19:41.117Z SwingSet: xsnap: v11: RangeError#1: Expected undefined is same as Interface
ag-power-tools-agd-1  | 2024-01-22T17:19:41.118Z SwingSet: xsnap: v11: RangeError: Expected (an undefined) is same as (a string)
ag-power-tools-agd-1  |  at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:243)
ag-power-tools-agd-1  |  at fail (/bundled-source/.../node_modules/ses/src/error/assert.js:357)
ag-power-tools-agd-1  |  at equal (/bundled-source/.../node_modules/ses/src/error/assert.js:390)
ag-power-tools-agd-1  |  at defendPrototype (/bundled-source/.../node_modules/@endo/exo/src/exo-tools.js:238)
ag-power-tools-agd-1  |  at defineKindInternal (/bundled-source/.../packages/swingset-liveslots/src/virtualObjectManager.js:957)
ag-power-tools-agd-1  |  at defineDurableKind (/bundled-source/.../packages/swingset-liveslots/src/virtualObjectManager.js:1155)
ag-power-tools-agd-1  |  at prepareAttenuator (.../internal/src/callback.js:300)
ag-power-tools-agd-1  |  at prepareGuardedAttenuator (.../internal/src/callback.js:319)
ag-power-tools-agd-1  |  at prepareMixinMyAddress (.../vats/src/nameHub.js:58)
ag-power-tools-agd-1  |  at prepareSpecializedNameAdmin (.../vats/src/vat-provisioning.js:20)
ag-power-tools-agd-1  |  at buildRootObject (.../vats/src/vat-provisioning.js:91)
```

</details>

### PR DESCRIPTION
**Pull request by mhofman**
closes: #8826 
refs: #8773
refs: #8802 #8793 and #7946

## Description

The release branch cherry-picks (with adaptations) the wallet factory and zoe+zcf  changes from master. However it also contains logic to perform the core proposal upgrade of these vats, which master lacks. This was previously blocked on #8826 which prevented upgrading vats when a new liveslots supervisor with a newer endo version with incompatible patterns packages is in use. This PR includes the fix provided by https://github.com/endojs/endo/pull/2038 as a patch to put master back in sync with the release branch.

### Security Considerations

None

### Scaling Considerations

None

### Documentation Considerations

None

### Testing Considerations

Integration ran manually and will need to keep passing for merging

### Upgrade Considerations

This replicates what the upgrade release branch does. Once the a3p base image is updated to capture the upcoming upgrade-14, the core proposals will need to be removed here (or at least the tests updated).

### CODE REVIEW

File: golang/cosmos/app/app.go [hunk 1]
```diff
@@ -845,6 +845,10 @@ func unreleasedUpgradeHandler(app *GaiaApp, targetUpgrade string) func(sdk.Conte
 		// Each CoreProposalStep runs sequentially, and can be constructed from
 		// one or more modules executing in parallel within the step.
 		CoreProposalSteps := []vm.CoreProposalStep{
+			// First, upgrade wallet factory
+			vm.CoreProposalStepForModules("@agoric/builders/scripts/smart-wallet/build-wallet-factory2-upgrade.js"),
+			// Then, upgrade Zoe and ZCF
+			vm.CoreProposalStepForModules("@agoric/builders/scripts/vats/replace-zoe.js"),
```
Comment (dckc) [inline golang/cosmos/app/app.go:851]: I'm only vaguely familiar with this `CoreProposalSteps` go stuff. I presume that the mechanism itself is sufficiently tested and the tests in`post.test.js` suffice to show that this usage is wired up right.

File: patches/@endo+patterns+1.1.0.patch [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+-  getInterfaceGuardPayload,
+   kindOf,
+ } from './src/patterns/patternMatchers.js';
+ 
++export {
++  getAwaitArgGuardPayload,
++  getMethodGuardPayload,
++  getInterfaceGuardPayload,
++  getInterfaceMethodKeys,
++} from './src/patterns/getGuardPayloads.js';
```
Comment (dckc) [inline patches/@endo+patterns+1.1.0.patch:38]: I'm not confident about how to review a patch in detail, but this part of it seems to correspond to
https://github.com/endojs/endo/pull/2038/files#diff-785fa700ea5ccbedf6df4f2d5f2abc2b68e4e7d87dd2c5638b264c49f1f90509

Comment (dckc) [reply patches/@endo+patterns+1.1.0.patch:38]: How do we usually track the work to get rid of such a patch when it's no longer needed?

**Review by dckc — APPROVED** [review]
I don't see any problems.

I'm not certain I know enough to approve this, but given the test coverage, I'll go ahead.

File: patches/@endo+patterns+1.1.0.patch (hunk 2 above)
Comment (mhofman) [reply patches/@endo+patterns+1.1.0.patch:38]: This is called out in the [maintainers steps](https://github.com/Agoric/agoric-sdk/blob/d4925683c81b6324fcc3c26ec7a6805ed193a9e0/MAINTAINERS.md?plain=1#L270-L274).

I also plan on staging a commit removing the patch in the [`integration-endo-master` branch, used for nightly integration with endo](https://github.com/Agoric/agoric-sdk/blob/d4925683c81b6324fcc3c26ec7a6805ed193a9e0/.github/actions/restore-node/action.yml#L87-L98).

File: golang/cosmos/app/app.go (hunk 1 above)
Comment (mhofman) [reply golang/cosmos/app/app.go:851]: This is the mechanism now used in the release branch, so this is just replicating that here on the master branch

File: patches/@endo+patterns+1.1.0.patch (hunk 2 above)
Comment (mhofman) [reply patches/@endo+patterns+1.1.0.patch:38]: removal staged here: https://github.com/Agoric/agoric-sdk/commit/0104d4d6a1e2973e336967db11a51993805f059f

**Pull request merged**